### PR TITLE
feat - add index study_id column on trials table

### PR DESCRIPTION
**Pull request by Ilevk**
## Motivation
I have experienced bottleneck in using rdbstorage with access more than 100 sessions simultaneously.
contribute #4444 

## Description of the changes

- add index to `study_id` on trials table

### CODE REVIEW

**Comment by c-bata** [comment]
Thank you for your pull request!
I have some questions and suggestions.

* Could you add a schema migration file using alembic?  See [wiki](https://github.com/optuna/optuna/wiki/RDB-schema-migration) for details.
* I guess a composite index of `study_id` and `state` columns might be more efficient. Is it possible to benchmark in your environment and share the result with us?

**Comment by Ilevk** [comment]
Okay, I'll try it and share it with you.

**Comment by Ilevk** [comment]
I tried composit index for 3 days and didn't see any dramatic performance changes. 

스크린샷 2023-02-25 오후 1 45 33

스크린샷 2023-02-25 오후 1 45 21

**Comment by Ilevk** [comment]
We found that migration does not work properly with the optuna storage upgrade command. However, the alembic upgrade head command works fine. Did I make a mistake?

스크린샷 2023-02-25 오후 2 34 05

**Comment by Ilevk** [comment]
I would expect it to be more efficient if I ran a large number of experiments per HPO. 
In my case, I run 10 experiments per HPO, which seems to make a small difference in performance.

**Review by c-bata — COMMENTED** [review]
Sorry for the late review. I'll review this PR today. Let me leave an early feedback for now.

File: optuna/storages/_rdb/alembic/versions/v3.1.0.a_.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+Revision ID: v3.1.0.a
+Revises: v3.0.0.d
+Create Date: 2023-02-25 13:21:00.730272
+
+"""
+from alembic import op
+
+
+# revision identifiers, used by Alembic.
+revision = "v3.1.0.a"
```
[the remark this answers is not in the document]
Comment (Ilevk) [reply optuna/storages/_rdb/alembic/versions/v3.1.0.a_.py]: @c-bata I've done to change revision and update db binary file

File: optuna/storages/_rdb/alembic/versions/v3.2.0.a_.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    op.create_index(op.f("trials_study_id_key"), "trials", ["study_id"], unique=False)
+
+
+def downgrade():
+    op.drop_index(op.f("trials_study_id_key"), table_name="trials")
```
Comment (c-bata) [inline optuna/storages/_rdb/alembic/versions/v3.2.0.a_.py:23]: Could you leave the following comment to let other developers know that this operation does not work with MySQL? Or you can simply skip the following operation when the sqlalchemy's dialect is mysql.

```suggestion
def downgrade():
    # Please note that the following operation does not work on MySQL due to a foreign key constraint.
    #
    # mysql> DROP INDEX ix_trials_study_id ON trials;
    # ERROR 1553 (HY000): Cannot drop index 'ix_trials_study_id': needed in a foreign key constraint
    op.drop_index(op.f("trials_study_id_key"), table_name="trials")
```

Comment (c-bata) [reply optuna/storages/_rdb/alembic/versions/v3.2.0.a_.py:23]: [FYI] Here is the error log when executing this `downgrade()` operation on MySQL.

```
$ cd optuna/storages/_rdb
$ git diff
diff --git a/optuna/storages/_rdb/alembic.ini b/optuna/storages/_rdb/alembic.ini
index 52681f91a..9311cded0 100644
--- a/optuna/storages/_rdb/alembic.ini
+++ b/optuna/storages/_rdb/alembic.ini
@@ -36,7 +36,7 @@ script_location = alembic
 # output_encoding = utf-8

 # NOTE: This URL is only used when generating migration scripts.
-sqlalchemy.url = sqlite:///alembic.db
+sqlalchemy.url = mysql+pymysql://optuna:password@127.0.0.1:3306/optuna
```

```
$ alembic downgrade v3.0.0.d
INFO  [alembic.runtime.migration] Context impl MySQLImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
INFO  [alembic.runtime.migration] Running downgrade v3.2.0.a -> v3.0.0.d, Add index to study_id column in trials table
Traceback (most recent call last):
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1900, in _execute_context
    self.dialect.do_execute(
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/sqlalchemy/engine/default.py", line 736, in do_execute
    cursor.execute(statement, parameters)
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/pymysql/cursors.py", line 148, in execute
    result = self._query(query)
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/pymysql/cursors.py", line 310, in _query
    conn.query(q)
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/pymysql/connections.py", line 548, in query
    self._affected_rows = self._read_query_result(unbuffered=unbuffered)
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/pymysql/connections.py", line 775, in _read_query_result
    result.read()
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/pymysql/connections.py", line 1156, in read
    first_packet = self.connection._read_packet()
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/pymysql/connections.py", line 725, in _read_packet
    packet.raise_for_error()
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/pymysql/protocol.py", line 221, in raise_for_error
    err.raise_mysql_exception(self._data)
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/pymysql/err.py", line 143, in raise_mysql_exception
    raise errorclass(errno, errval)
pymysql.err.OperationalError: (1091, "Can't DROP 'trials_study_id_key'; check that column/key exists")

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/bin/alembic", line 8, in <module>
    sys.exit(main())
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/alembic/config.py", line 588, in main
    CommandLine(prog=prog).main(argv=argv)
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/alembic/config.py", line 582, in main
    self.run_cmd(cfg, options)
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/alembic/config.py", line 559, in run_cmd
    fn(
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/alembic/command.py", line 366, in downgrade
    script.run_env()
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/alembic/script/base.py", line 563, in run_env
    util.load_python_file(self.dir, "env.py")
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/alembic/util/pyfiles.py", line 92, in load_python_file
    module = load_module_py(module_id, path)
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/alembic/util/pyfiles.py", line 108, in load_module_py
    spec.loader.exec_module(module)  # type: ignore
  File "<frozen importlib._bootstrap_external>", line 850, in exec_module
  File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
  File "alembic/env.py", line 79, in <module>
    run_migrations_online()
  File "alembic/env.py", line 73, in run_migrations_online
    context.run_migrations()
  File "<string>", line 8, in run_migrations
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/alembic/runtime/environment.py", line 851, in run_migrations
    self.get_context().run_migrations(**kw)
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/alembic/runtime/migration.py", line 620, in run_migrations
    step.migration_fn(**kw)
  File "/Users/c-bata/go/src/github.com/optuna/optuna/optuna/storages/_rdb/alembic/versions/v3.2.0.a_.py", line 23, in downgrade
    op.drop_index(op.f("trials_study_id_key"), table_name="trials")
  File "<string>", line 8, in drop_index
  File "<string>", line 3, in drop_index
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/alembic/operations/ops.py", line 1077, in drop_index
    return operations.invoke(op)
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/alembic/operations/base.py", line 392, in invoke
    return fn(self, operation)
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/alembic/operations/toimpl.py", line 104, in drop_index
    operations.impl.drop_index(
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/alembic/ddl/impl.py", line 391, in drop_index
    self._exec(schema.DropIndex(index))
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/alembic/ddl/impl.py", line 197, in _exec
    return conn.execute(construct, multiparams)
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1380, in execute
    return meth(self, multiparams, params, _EMPTY_EXECUTION_OPTS)
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/sqlalchemy/sql/ddl.py", line 80, in _execute_on_connection
    return connection._execute_ddl(
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1472, in _execute_ddl
    ret = self._execute_context(
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1943, in _execute_context
    self._handle_dbapi_exception(
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 2124, in _handle_dbapi_exception
    util.raise_(
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/sqlalchemy/util/compat.py", line 211, in raise_
    raise exception
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1900, in _execute_context
    self.dialect.do_execute(
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/sqlalchemy/engine/default.py", line 736, in do_execute
    cursor.execute(statement, parameters)
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/pymysql/cursors.py", line 148, in execute
    result = self._query(query)
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/pymysql/cursors.py", line 310, in _query
    conn.query(q)
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/pymysql/connections.py", line 548, in query
    self._affected_rows = self._read_query_result(unbuffered=unbuffered)
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/pymysql/connections.py", line 775, in _read_query_result
    result.read()
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/pymysql/connections.py", line 1156, in read
    first_packet = self.connection._read_packet()
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/pymysql/connections.py", line 725, in _read_packet
    packet.raise_for_error()
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/pymysql/protocol.py", line 221, in raise_for_error
    err.raise_mysql_exception(self._data)
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/pymysql/err.py", line 143, in raise_mysql_exception
    raise errorclass(errno, errval)
sqlalchemy.exc.OperationalError: (pymysql.err.OperationalError) (1091, "Can't DROP 'trials_study_id_key'; check that column/key exists")
[SQL:
DROP INDEX trials_study_id_key ON trials]
(Background on this error at: https://sqlalche.me/e/14/e3q8)
```

**Review by c-bata — COMMENTED** [review]
Thank you for the quick update! I left a minor suggestion.

Regarding the migration script, it looks good to me. The backward migration does not work in MySQL, but is acceptable since Optuna does not actually provide an API for backward migration to users.

I am checking for performance gains with this change using the following script.
https://gist.github.com/c-bata/c08fb89a583adbcdc3eddcf8cf192c1a

* In MySQL, the performance improvement could not be found because the index already exists due to foreign key constraints.
* In SQLite3, as you can see the following comment, no significant difference seems to exist with 50 studies (100 trials per each).
https://gist.github.com/c-bata/c08fb89a583adbcdc3eddcf8cf192c1a?permalink_comment_id=4500763#gistcomment-4500763

I will verify the performance on PostgreSQL tomorrow with more study and trial records.

**Comment by c-bata** [comment]
I have confirmed with PostgreSQL that this change improves performance. I will approve this PR after my suggestion is reflected 👍 

Here is a benchmark script and its result.

Benchmark script:

https://gist.github.com/c-bata/98532a60609a8a5f9e1e4dd162d45886

Before (master):

```
$ python profile_get_all_trials.py
Elapsed: 24.4298s (n_trials=500 n_params=10)

Sort by Total:
Total Time(s)	Query Count	Statement
0.5289	100	SELECT trial_params.trial_id AS trial_params_trial_id, trial_params.param_id AS trial_params_param_id, trial_params.param_name AS trial_params_param_name, trial_params.param_value AS trial_params_param_value, trial_params.distribution_json AS trial_params_distribution_json
FROM trial_params
WHERE trial_params.trial_id IN (%(primary_keys_1)s, %(primary_keys_2)s, %(primary_keys_3)s, %(primary_keys_4)s, %(primary_keys_5)s, ...)
0.2383	100	SELECT trials.trial_id AS trials_trial_id
FROM trials
WHERE trials.study_id = %(study_id_1)s
```

After (This PR):

```
# python profile_all_trials.py
Elapsed: 21.5016s (n_trials=500 n_params=10)

Sort by Total:
Total Time(s)	Query Count	Statement
0.4890	100	SELECT trial_params.trial_id AS trial_params_trial_id, trial_params.param_id AS trial_params_param_id, trial_params.param_name AS trial_params_param_name, trial_params.param_value AS trial_params_param_value, trial_params.distribution_json AS trial_params_distribution_json
FROM trial_params
WHERE trial_params.trial_id IN (%(primary_keys_1)s, %(primary_keys_2)s, %(primary_keys_3)s, %(primary_keys_4)s, ...)
0.2103	100	SELECT trials.trial_id AS trials_trial_id, trials.number AS trials_number, trials.study_id AS trials_study_id, trials.state AS trials_state, trials.datetime_start AS trials_datetime_start, trials.datetime_complete AS trials_datetime_complete
FROM trials
WHERE trials.trial_id IN (%(trial_id_1_1)s, %(trial_id_1_2)s, %(trial_id_1_3)s, %(trial_id_1_4)s, ...) AND trials.study_id = %(study_id_1)s ORDER BY trials.trial_id
```

According to the slow queries, a composite index of `study_id` and `state` does not improves the performance since `state` is not included in the query.

**Comment by toshihikoyanase** [comment]
@Alnusjaponica Could you review this PR, please?
Let me explain how to confirm the schema migration with the [optuna-e2e](https://github.com/himkt/optuna-e2e) tool together.

**Comment by Alnusjaponica** [comment]
Thank you for your contribution.
I also confirmed that migration works properly and the changes looks good to me.

**Comment by Alnusjaponica** [comment]
@Ilevk Thank you for your contribution.
I also confirmed that the migration code works properly and the change itself looks good to me.

As we discussed offline, @c-bata found out that this change might not affect the performance (he'll share the data presently) and wonders what made this change affect the performance drastically in your environment. Could you provide us some information about what kind of job you're running or any reproducible codes?

**Comment by Ilevk** [comment]
@Alnusjaponica We train about 5,000 models every day, and we run 5-7 HPOs per model. We have about 100 training instances. If you have a lot of connections happening at the same time, you seem to have the same problem as me.

**Comment by c-bata** [comment]
@Ilevk Thank you for your swift response. Could you also share which sampler you used and the number of trials per study?

**Comment by Ilevk** [comment]
@c-bata we use a TPESampler & 5 ~ 7 trials each. almost 5 trials.

**Comment by Alnusjaponica** [comment]
Thank you for the information! We will run the benchmark again to investigate the situation.

**Comment by Ilevk** [comment]
There is one more bottleneck in our environment. Currently, RDBStorage creates engines internally with create_engine, and when accessed by many instances at the same time, the previously created engines & connections are not cleaned up and delay the next experiment.

**Comment by Ilevk** [comment]
Our team modified the code to reuse the connection engine.

**Review by c-bata — APPROVED** [review]
LGTM! I could see a clear performance improvement in the following scenario.

https://gist.github.com/c-bata/98532a60609a8a5f9e1e4dd162d45886

## Before

```
optuna=# \d trials;
                                               Table "public.trials"
      Column       |            Type             | Collation | Nullable |                 Default
-------------------+-----------------------------+-----------+----------+------------------------------------------
 trial_id          | integer                     |           | not null | nextval('trials_trial_id_seq'::regclass)
 number            | integer                     |           |          |
 study_id          | integer                     |           |          |
 state             | trialstate                  |           | not null |
 datetime_start    | timestamp without time zone |           |          |
 datetime_complete | timestamp without time zone |           |          |
Indexes:
    "trials_pkey" PRIMARY KEY, btree (trial_id)
Foreign-key constraints:
    "trials_study_id_fkey" FOREIGN KEY (study_id) REFERENCES studies(study_id)
Referenced by:
    TABLE "trial_heartbeats" CONSTRAINT "trial_heartbeats_trial_id_fkey" FOREIGN KEY (trial_id) REFERENCES trials(trial_id)
    TABLE "trial_intermediate_values" CONSTRAINT "trial_intermediate_values_trial_id_fkey" FOREIGN KEY (trial_id) REFERENCES trials(trial_id)
    TABLE "trial_params" CONSTRAINT "trial_params_trial_id_fkey" FOREIGN KEY (trial_id) REFERENCES trials(trial_id)
    TABLE "trial_system_attributes" CONSTRAINT "trial_system_attributes_trial_id_fkey" FOREIGN KEY (trial_id) REFERENCES trials(trial_id)
    TABLE "trial_user_attributes" CONSTRAINT "trial_user_attributes_trial_id_fkey" FOREIGN KEY (trial_id) REFERENCES trials(trial_id)
    TABLE "trial_values" CONSTRAINT "trial_values_trial_id_fkey" FOREIGN KEY (trial_id) REFERENCES trials(trial_id)
```

```
# python profiler.py
Elapsed: 157.2042s (n_trials=500 n_params=10)

Sort by Total:
Total Time(s)	Query Count	Statement
43.5951	11000	SELECT trials.trial_id AS trials_trial_id
FROM trials
WHERE trials.study_id = %(study_id_1)s
40.9885	10000	SELECT trial_params.param_id AS trial_params_param_id, trial_params.trial_id AS trial_params_trial_id, trial_params.param_name AS trial_params_param_name, trial_params.param_value AS trial_params_param_value, trial_params.distribution_json AS trial_params_distribution_json
FROM trial_params JOIN trials ON trials.trial_id = trial_params.trial_id
WHERE trials.study_id = %(study_id_1)s AND trial_params.param_name = %(param_name_1)s
 LIMIT %(param_1)s
4.7336	21000	SELECT trials.trial_id AS trials_trial_id, trials.number AS trials_number, trials.study_id AS trials_study_id, trials.state AS trials_state, trials.datetime_start AS trials_datetime_start, trials.datetime_complete AS trials_datetime_complete
FROM trials
WHERE trials.trial_id = %(trial_id_1)s
4.5883	1000	SELECT count(trials.trial_id) AS count_1
FROM trials
WHERE trials.study_id = %(study_id_1)s AND trials.trial_id < %(trial_id_1)s
3.5560	11100	SELECT studies.study_id AS studies_study_id, studies.study_name AS studies_study_name
FROM studies
WHERE studies.study_id = %(study_id_1)s
```

## After

```
optuna=# create index ix_trials_study_id on trials(study_id);
CREATE INDEX
optuna=# \d trials;
                                               Table "public.trials"
      Column       |            Type             | Collation | Nullable |                 Default
-------------------+-----------------------------+-----------+----------+------------------------------------------
 trial_id          | integer                     |           | not null | nextval('trials_trial_id_seq'::regclass)
 number            | integer                     |           |          |
 study_id          | integer                     |           |          |
 state             | trialstate                  |           | not null |
 datetime_start    | timestamp without time zone |           |          |
 datetime_complete | timestamp without time zone |           |          |
Indexes:
    "trials_pkey" PRIMARY KEY, btree (trial_id)
    "ix_trials_study_id" btree (study_id)
Foreign-key constraints:
    "trials_study_id_fkey" FOREIGN KEY (study_id) REFERENCES studies(study_id)
Referenced by:
    TABLE "trial_heartbeats" CONSTRAINT "trial_heartbeats_trial_id_fkey" FOREIGN KEY (trial_id) REFERENCES trials(trial_id)
    TABLE "trial_intermediate_values" CONSTRAINT "trial_intermediate_values_trial_id_fkey" FOREIGN KEY (trial_id) REFERENCES trials(trial_id)
    TABLE "trial_params" CONSTRAINT "trial_params_trial_id_fkey" FOREIGN KEY (trial_id) REFERENCES trials(trial_id)
    TABLE "trial_system_attributes" CONSTRAINT "trial_system_attributes_trial_id_fkey" FOREIGN KEY (trial_id) REFERENCES trials(trial_id)
    TABLE "trial_user_attributes" CONSTRAINT "trial_user_attributes_trial_id_fkey" FOREIGN KEY (trial_id) REFERENCES trials(trial_id)
    TABLE "trial_values" CONSTRAINT "trial_values_trial_id_fkey" FOREIGN KEY (trial_id) REFERENCES trials(trial_id)
```

```
# python profiler.py
Elapsed: 66.0051s (n_trials=500 n_params=10)

Sort by Total:
Total Time(s)	Query Count	Statement
4.3927	21000	SELECT trials.trial_id AS trials_trial_id, trials.number AS trials_number, trials.study_id AS trials_study_id, trials.state AS trials_state, trials.datetime_start AS trials_datetime_start, trials.datetime_complete AS trials_datetime_complete
FROM trials
WHERE trials.trial_id = %(trial_id_1)s
3.0978	10000	SELECT trial_params.param_id AS trial_params_param_id, trial_params.trial_id AS trial_params_trial_id, trial_params.param_name AS trial_params_param_name, trial_params.param_value AS trial_params_param_value, trial_params.distribution_json AS trial_params_distribution_json
FROM trial_params JOIN trials ON trials.trial_id = trial_params.trial_id
WHERE trials.study_id = %(study_id_1)s AND trial_params.param_name = %(param_name_1)s
 LIMIT %(param_1)s
2.9455	11100	SELECT studies.study_id AS studies_study_id, studies.study_name AS studies_study_name
FROM studies
WHERE studies.study_id = %(study_id_1)s
2.3931	10000	INSERT INTO trial_params (trial_id, param_name, param_value, distribution_json) VALUES (%(trial_id)s, %(param_name)s, %(param_value)s, %(distribution_json)s) RETURNING trial_params.param_id
1.8892	11000	SELECT trials.trial_id AS trials_trial_id, trials.number AS trials_number, trials.study_id AS trials_study_id, trials.state AS trials_state, trials.datetime_start AS trials_datetime_start, trials.datetime_complete AS trials_datetime_complete
FROM trials
WHERE trials.trial_id IN (NULL) AND (1 != 1) AND trials.study_id = %(study_id_1)s ORDER BY trials.trial_id
```

**Comment by c-bata** [comment]
> There is one more bottleneck in our environment. Currently, RDBStorage creates engines internally with create_engine, and when accessed by many instances at the same time, the previously created engines & connections are not cleaned up and delay the next experiment.

@Ilevk  Thank you for sharing. The connection objects are basically cleaned up when the reference count of RDBStorage reaches zero, but if there are any connections left, they can be explicitly cleaned up as follows.

```python
storage = optuna.storages.RDBStorage(storage_url)
study = optuna.create_study(storage=storage)
study.optimize(objective, ...)

# Explicitly clean up connections
storage.engine.dispose()
```

If you find any problems with the handling of connection objects in Optuna, please report them to us.

**Comment by Alnusjaponica** [comment]
Sorry for my delayed reply. I also confirmed that the newer version is about twice  as fast as the older version.

**Review by Alnusjaponica — APPROVED** [review]
Sorry for my delayed reply. I also run optimizations in the same scenario with two different versions and confirmed that it is the newer one is about twice as fast as the older one. LGTM.

**Pull request merged**